image-hd.c

```diff
@@ -645,28 +739,37 @@ static int hdimage_setup(struct image *image, cfg_t *cfg)
 					part->name);
 			return -EINVAL;
 		}
+		if (!part->extended) {
+			int ret = check_overlap(image, part);
+			if (ret)
+				return ret;
+		} else if (now > part->offset) {
+			image_error(image, "part %s overlaps with previous partition\n",
+				    part->name);
+			return -EINVAL;
+		}
 		if (part->in_partition_table && (part->size % 512)) {
 			image_error(image, "part %s size (%lld) must be a "
 					"multiple of 1 sector (512 bytes)\n",
 					part->name, part->size);
 			return -EINVAL;
 		}
-		now = part->offset + part->size;
+		if (part->offset + part->size > now)
+			now = part->offset + part->size;
 	}
 
-	if (hd->gpt)
-		now += GPT_SECTORS * 512;
-
 	if (image->size > 0 && now > image->size) {
 		image_error(image, "partitions exceed device size\n");
 		return -EINVAL;
 	}
 
 	if (image->size == 0) {
-		if (hd->gpt)
+		if (hd->gpt) {
+			now += GPT_SECTORS * 512;
```

Comment:
In the case where the image size is not given in advance, I do need to make room for the backup table, which is why that increment instead happens in here instead. One could argue whether one should actually do this when no-gpt-backup is set, but it's size was accounted for previously (both with and without a given image size), so AFAICT I'm not changing behaviour in that regard. But it's possible that I shouldn't add the "GPT backup" fake partition when no-gpt-backup.

Reply:
The space for the gpt backup should always be reserved. The use-case is, to keep the image smaller and do a fixup at runtime. So there needs to be some space left to actually write the table.